### PR TITLE
General refactor

### LINKED ISSUE #3: DOM effect
Should `localStorage` and `sessionStorage` have the type `Eff (storage :: STORAGE, dom :: DOM | e) Storage` since they're dependent on `Window`?

### PR DESCRIPTION
**Pull request by texastoland**
TODO:
- [x] #3 `DOM` effect
- [x] #5 Non-generic transcoding
- [x] #2 Mock
- [x] Tests (using Spec)
- [x] #7 Replace records with <s>classes</s> named functions
- [x] <s>#8 Operators</s> (using `Var`)
- [x] `package.json` (copied from Aff)
- [x] CI (copied from Aff)
- [ ] Documentation _(help wanted)_
- [ ] Move to @purescript-contrib or @purescripters?

Also:
- ignore IDE files
- implicit dependencies (Argonaut pulls in the majority)
- <s>newtype for `Generic` transcoding</s> resulted in purescript-contrib/purescript-argonaut-codecs#20 (`JSON.purs` is copy pasta from `Generic.purs` since there's no way to generalize constraints)
- contrib's coding conventions

### CODE REVIEW

**Comment by texastoland** [comment]
Hmm GitHub diff isn't quite as meticulous as local. I kept more than the diff reveals.

**Comment by texastoland** [comment]
Ready for review 😪

**Comment by texastoland** [comment]
CI passing 🎆

**Comment by eskimor** [comment]
Woah - ok someone did some work. Review in progress ...

**Comment by eskimor** [comment]
might take a bit - I am actually on holiday right now ;-)

**Comment by texastoland** [comment]
No worries! I reviewed myself by opening old and new branches side by side since the diffs are terrible.

File: src/DOM/WebStorage.purs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  , getItem
+  , length
+  , removeItem
+  , setItem
+  , updateItem
+  , updateItem'
+  ) as Exports
+import DOM.WebStorage.Local (LocalStorage, getLocalStorage) as Exports
+import DOM.WebStorage.Mock (MockStorage, newMockStorage) as Exports
+import DOM.WebStorage.Session (SessionStorage, getSessionStorage) as Exports
```
Comment (eskimor) [inline src/DOM/WebStorage.purs]: I like :+

File: src/DOM/WebStorage/Internal/Generic.purs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  compare = gCompare
+
+instance encodeTranscodeG :: Generic a => EncodeJson (TranscodeG a) where
+  encodeJson = gEncodeJson
+
+instance decodeTranscodeG :: Generic a => DecodeJson (TranscodeG a) where
+  decodeJson = gDecodeJson
+
+unTranscodeG :: forall a. TranscodeG a -> a
+unTranscodeG (TranscodeG item) = item
```
Comment (eskimor) [inline src/DOM/WebStorage/Internal/Generic.purs]: Neat trick! But I think I'd like a separate type class for generic encoding better - this would make user code as clear as before again. (No TranscodeG distraction.) In other words, in addition to JSONStorage a GenericStorage class in a file Generic.purs, with the same functions, but Generic constraint instead of the ToJSON/FromJSON constraints.

Comment (texastoland) [inline src/DOM/WebStorage/Internal/Generic.purs]: Hmm that's brilliant. I could actually create an instance that extends the coders one just like `JSONStorage` provides an implicit instance for `StrStorage` now.

File: src/DOM/WebStorage/Local.purs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+newtype LocalStorage = LocalStorage ForeignStorage
+
+instance strStorageLocalStorage :: StrStorage LocalStorage where
+  length = length <<< unLocalStorage
+  key = key <<< unLocalStorage
+  getItem = getItem <<< unLocalStorage
+  setItem = setItem <<< unLocalStorage
+  removeItem = removeItem <<< unLocalStorage
+  clear = clear <<< unLocalStorage
+
```
Comment (eskimor) [inline src/DOM/WebStorage/Local.purs]: Why is this necessary? Can't localStorage and sessionStorage just be of type ForeignStorage? What benefit has the distinction?

Comment (texastoland) [inline src/DOM/WebStorage/Local.purs]: They have to have a `newtype` so their instances don't collide.

File: src/DOM/WebStorage/Internal/Generic.purs (hunk 2 above)
Comment (eskimor) [inline src/DOM/WebStorage/Internal/Generic.purs]: hmm, I am not really able to follow ..

File: src/DOM/WebStorage/Local.purs (hunk 3 above)
Comment (eskimor) [inline src/DOM/WebStorage/Local.purs]: So PS does not allow instances on foreign types? Ok, but still sessionStorage and localStorage could still share the same StrStorage instance -right? and therefore could share the same newtype around ForeignStorage? Same for MockStorage actually, as you made the JavaScript API a ForeignStorage too there. They are all the same type, I don't really get why you need to make them different with newtypes. What do I miss?

Comment (texastoland) [inline src/DOM/WebStorage/Local.purs]: Ah you're right 🙈

Comment (texastoland) [inline src/DOM/WebStorage/Local.purs]: Fixed 👌

File: src/DOM/WebStorage/Internal/Generic.purs (hunk 2 above)
Comment (texastoland) [inline src/DOM/WebStorage/Internal/Generic.purs]: Fixed 👌

File: src/DOM/WebStorage/Generic.purs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  length :: forall e. s -> Eff (storage :: STORAGE | e) Int
+  -- key would require gRead
+  getItem :: forall e key a. (Generic (key a), Generic a)
+    => s -> key a -> Eff (storage :: STORAGE | e) (Maybe a)
+  setItem :: forall e key a. (Generic (key a), Generic a)
+    => s -> key a -> a -> Eff (storage :: STORAGE | e) Unit
+  removeItem :: forall e key a. Generic (key a)
+    => s -> key a -> Eff (storage :: STORAGE | e) Unit
+  clear :: forall e. s -> Eff (storage :: STORAGE | e) Unit
+
```
Comment (texastoland) [inline src/DOM/WebStorage/Generic.purs]: Everything below has to be copy pasta without fancier types right CC @paf31 (see `class JSONStorage`)?

File: src/DOM/WebStorage.purs (hunk 1 above)
Comment (texastoland) [inline src/DOM/WebStorage.purs]: @garyb showed me that trick 🙌

Comment (texastoland) [inline src/DOM/WebStorage.purs]: @gb_r showed me that trick 🙌

**Comment by eskimor** [comment]
looks really good to me now! - I'll check more thoroughly & merge soon.

**Comment by texastoland** [comment]
About to push a fix that makes all instances `:: *Storage ForeignStorage`. That probably means the functions don't need to be methods anymore since the first term isn't polymorphic (originally needed by newtypes).

**Comment by texastoland** [comment]
It's much closer now to how it was originally which I like. One last idea: add `getItemRef` returning an actual `Ref` via FFI. `readRef` returns a value instead of `Maybe`. Perhaps `getItemRef` could accept a default like `maybe`.The benefits are operators (arriving in `2.0.0`) and passing like a global variable (it is) without the key.

**Comment by texastoland** [comment]
Deleting `updateItem` functions in favor of `modifyRef` (it's copy pasta from my PR there anyway). Will update tests this weekend!

**Comment by texastoland** [comment]
I showed @paf31 `getItemRef` and he pointed me to @zudov's [`Var`](https://github.com/zudov/purescript-var/blob/master/src/Control/Monad/Eff/Var.purs). I think it'd be a bit cleaner and a lot more usable if there's no objection.

**Comment by eskimor** [comment]
@texastoland Really nice work - I love how simple it got in the end - awesome! Even more awesome that we have a test suit now!

**Comment by eskimor** [comment]
> I showed @paf31 getItemRef and he pointed me to @zudov's Var. I think it'd be a bit cleaner and a lot more usable if there's no objection.

That could be nice actually!

``` purescript
 getLocalStorageRef myKey $= myvalue
```

instead of

``` purescript
  setItem getLocalStorage myKey myvalue
```

looks a tad nicer. Personally I would not bother, but I don't object either. (As additional feature - I assume?)

**Comment by texastoland** [comment]
> That could be nice actually! `getLocalStorageRef myKey $= myvalue`

https://github.com/eskimor/purescript-localstorage/blob/master/test/DOM/WebStorage/Var.purs#L36

> Personally I would not bother, but I don't object either.

The bigger selling points for me were an implementation of `update` (which I had previously pasted directly from [my `Ref` PR](https://github.com/texastoland/purescript-refs/blob/bb6403c053caed4f408df2d73e1f2bbb16326469/src/Control/Monad/Eff/Ref.purs#L38)) and a number of instances (including `Functor`) for free.

> - [ ] Documentation _(help wanted)_

I edited `Example.purs` but not the `README`. I think it'd be enough to mention the former in the latter. I can also help with doc comments eventually.

> - [ ] Move to @purescript-contrib or @purescripters?

Does this sound appealing? Just an idea for discovery. It might also make sense to name `webstorage` (after the W3C spec) but neither is crucial.

**Comment by eskimor** [comment]
I did not choose purescript-webstorage, because there was already a package named like [that](https://github.com/joneshf/purescript-webstorage). We could move it to purescript-contrib, no problem with that - what do I have to do for this to happen?

I am going to update the readme, then I am going to throw out a new release? Or do you want something else in before?

**Comment by texastoland** [comment]
@joneshf's (I type @st58 every time) was never published but it does look like https://pursuit.purescript.org/packages/purescript-web-storage/1.0.0 took a similar name. I assume all that's necessary for contrib is @paf31 to add you? Other than that I think it's release ready!

**Comment by joneshf** [comment]
`purescript-webstorage` has been published for more than two years. It's not really being actively developed anymore though, so I don't want to name squat. However, there's a couple of libraries now, so I'm not sure what to do there.

**Comment by texastoland** [comment]
> `purescript-webstorage` has been published for more than two years.

My bad @joneshf. It must not have gotten synced one of the times Pursuit broke. I was basing it off joneshf/purescript-webstorage#5. FWIW I referenced your implementation again this refactor.

> However, there's a couple of libraries now, so I'm not sure what to do there.

Perhaps @ianbollinger would be interested in a PR based on https://github.com/eskimor/purescript-localstorage/blob/master/src/DOM/WebStorage/String.purs? That way one could provide fairly naive bindings and the other could add the JSON and phantom types magic.

**Comment by ianbollinger** [comment]
I probably should have taken a closer look at other libraries before writing mine, oops! I couldn't find what I needed on Pursuit. Anyway, making the APIs of the two libraries align would be fine with me (currently, the parameter order is different). Or if you want to combine projects, that would also be fine. Though having a lower-level version that doesn't rely on `purescript-dom` might be desirable, since it's pretty big.

Also, [setItem](https://github.com/eskimor/purescript-localstorage/blob/master/src/DOM/WebStorage/String.purs#L33) can throw an exception. Actually, according to the spec, just accessing the localStorage attribute can throw an exception too, which I neglected to account for in my API.

**Comment by joneshf** [comment]
Ah, yeah pursuit didn't exist when it was being actively developed, so that's why it's not there. :smile: It's on bower though.

**Comment by texastoland** [comment]
> making the APIs of the two libraries align would be fine with me

I'll try to PR this weekend!

> (currently, the parameter order is different).

Intentional:
1. to mirror `Ref` and `STRef`
2. and it's less useful…
   - to curry a specific key with different storages
   - than vice versa e.g. [`getItemVar`](https://github.com/eskimor/purescript-localstorage/blob/master/src/DOM/WebStorage/String.purs#L45-L50)

>  Or if you want to combine projects, that would also be fine.

IMO there's a case for modularity. @eskimor?

> Though having a lower-level version that doesn't rely on `purescript-dom` might be desirable, since it's pretty big.

See https://github.com/eskimor/purescript-localstorage/issues/3#issuecomment-236645924.

> Also, `setItem` can throw an exception.

Good catch I'll add it!

> Actually, according to the spec, just accessing the `localStorage` attribute can throw an exception too,

Hmm [you're correct](http://www.w3.org/TR/webstorage/#dom-localstorage) but it's curious because IIUC browsers <q>do nothing</q> instead of error in practice.

**Comment by eskimor** [comment]
I would not split it personally, but it does not matter for me - so if a split seems to be useful for one of you - it is fine with me!
